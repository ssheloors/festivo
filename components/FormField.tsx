import { ReactNode } from "react";
import {
  Controller,
  ControllerProps,
  FieldError,
  FieldValues,
  Path,
  UseFormReturn,
} from "react-hook-form";
import { NativeSyntheticEvent, NativeTouchEvent } from "react-native";
import {
  InputProps,
  Label,
  SizableText,
  SizableTextProps,
  YStack,
} from "tamagui";

import { Input } from "./Input";

export function FormField<T extends FieldValues>({
  form,
  label,
  name,
  inputProps,
  render,
}: {
  form: UseFormReturn<T>;
  label: ReactNode;
  name: Path<T>;
  inputProps?: InputProps;
  render?: (
    props: {
      inputProps: {
        id: string;
        onPress: (e: NativeSyntheticEvent<NativeTouchEvent>) => void;
        opacity: number;
      };
    } & Parameters<ControllerProps["render"]>[0],
  ) => JSX.Element;
}) {
  return (
    <YStack alignItems="stretch" width="100%">
      <Label htmlFor={name}>{label}</Label>
      <Controller
        control={form.control}
        name={name}
        render={({ field: { onChange, ...field }, ...renderProps }) =>
          render ? (
            render({
              inputProps: {
                id: name,
                onPress: (e) => {
                  e.stopPropagation();
                },
                opacity: field.disabled ? 0.5 : 1,
              },
              field: { ...field, onChange },
              ...renderProps,
            })
          ) : (
            <Input
              id={name}
              onPress={(e) => {
                e.stopPropagation();
              }}
              opacity={field.disabled ? 0.5 : 1}
              {...inputProps}
              {...field}
              onChangeText={onChange}
            />
          )
        }
      />
      <ErrorMessage
        error={form.formState.errors[name] as FieldError | undefined}
      />
    </YStack>
  );
}

export function ErrorMessage({
  error,
  ...props
}: SizableTextProps & { error: FieldError | Error | undefined | null }) {
  if (!error) {
    return null;
  }

  return (
    <SizableText color="red" paddingBlock="$2" size="$2" {...props}>
      {error.message}
    </SizableText>
  );
}
