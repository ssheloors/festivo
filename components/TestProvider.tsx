import { TamaguiProvider } from "tamagui";

import tamaguiConfig from "@/tamagui.config";

export function TestProvider({ children }: { children: JSX.Element }) {
  return <TamaguiProvider config={tamaguiConfig}>{children}</TamaguiProvider>;
}
