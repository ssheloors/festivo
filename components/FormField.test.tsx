import {
  FieldValues,
  useForm,
  UseFormProps,
  UseFormReturn,
} from "react-hook-form";
import renderer from "react-test-renderer";

import { FormField } from "./FormField";
import { TestProvider } from "./TestProvider";

function FormWrapper<T extends FieldValues>({
  defaultValues,
  children,
}: {
  defaultValues?: UseFormProps<T>["defaultValues"];
  children: (form: UseFormReturn<T>) => JSX.Element;
}) {
  const form = useForm({ defaultValues });

  return children(form);
}

it(`renders correctly`, () => {
  const tree = renderer
    .create(
      <TestProvider>
        <FormWrapper defaultValues={{ testInput: "testInputValue" }}>
          {(form) => (
            <FormField
              form={form}
              label="Test input"
              name="testInput"
              inputProps={{ className: "test-class" }}
            />
          )}
        </FormWrapper>
      </TestProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
