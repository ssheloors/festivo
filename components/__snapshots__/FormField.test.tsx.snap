// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <span
    class=""
    style="display: contents;"
  >
    <span
      class=" _dsp_contents"
    >
      <span
        class=" t_light _dsp_contents is_Theme"
      >
        <div
          class="_dsp-flex _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column _ai-stretch"
        >
          <label
            class="is_Label font_body _col-0active-colorPress _dsp-flex _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _ff-f-family _fow-f-weight-4 _ls-f-letterSpa1360334202 _fos-f-size-4 _lh-t-size-true _col-color _bg-transparent _ai-center _ussel-none _cur-default"
            data-disable-theme="true"
            for="testInput"
            id=":r0:"
          >
            Test input
          </label>
          <span
            class=" _dsp_contents"
          >
            <span
              class="t_sub_theme t_Input _dsp_contents is_Theme"
            >
              <span
                class="_dsp_contents"
              >
                <input
                  aria-labelledby=":r0:"
                  autocapitalize="sentences"
                  autocomplete="on"
                  autocorrect="on"
                  class="css-textinput-11aywtz r-placeholderTextColor-6taxm2 is_Input font_body _btc-0hover-borderColor69916956 _brc-0hover-borderColor69916956 _bbc-0hover-borderColor69916956 _blc-0hover-borderColor69916956 _btc-0focus-borderColor68052152 _brc-0focus-borderColor68052152 _bbc-0focus-borderColor68052152 _blc-0focus-borderColor68052152 _outlineColor-0focus-visible-outlineColo114 _outlineWidth-0focus-visible-2px _outlineStyle-0focus-visible-solid _ff-f-family _fow-f-weight-4 _ls-f-letterSpa1360334202 _fos-f-size-4 _lh-f-lineHeigh112923 _pr-t-space-3--53 _pl-t-space-3--53 _h-t-size-true _btlr-t-radius-tr3728 _btrr-t-radius-tr3728 _bbrr-t-radius-tr3728 _bblr-t-radius-tr3728 _btw-1px _brw-1px _bbw-1px _blw-1px _outlineWidth-0px _col-color _btc-borderColor _brc-borderColor _bbc-borderColor _blc-borderColor _bg-background _miw-0px _bbs-solid _bts-solid _bls-solid _brs-solid _o-1 test-class"
                  data-disable-theme="true"
                  dir="auto"
                  id="testInput"
                  rows="1"
                  spellcheck="true"
                  style="--placeholderTextColor: var(--placeholderColor);"
                  tabindex="0"
                  value="testInputValue"
                  virtualkeyboardpolicy="auto"
                />
              </span>
            </span>
          </span>
        </div>
        <div
          style="display: contents;"
        />
      </span>
    </span>
  </span>
</DocumentFragment>
`;
